### ISSUE TITLE
format() is a foot-gun when used with pointers (segfault)

### ISSUE THREAD

**Issue by cztomsik** [ziglang/zig#15335]
### Zig Version

0.11.0-dev.2317+46b2f1f70

### Steps to Reproduce and Observed Behavior

Here's repro - I hope I haven't messed up anything else, because it was originally in a different code.

```
const std = @import("std");

const Animal = extern struct {
    kind: enum(usize) { cat, dog },

    pub fn format(self: *const Animal, comptime _: []const u8, _: std.fmt.FormatOptions, writer: anytype) !void {
        switch (self.kind) {
            .cat => @panic("not implemented"),
            .dog => try writer.print("{}", .{@ptrCast(*const Dog, self)}),
        }
    }
};

const Dog = extern struct {
    base: Animal,
    name: [*:0]const u8,

    pub fn format(self: *const Dog, comptime _: []const u8, _: std.fmt.FormatOptions, writer: anytype) !void {
        try writer.print("Dog {s}", .{self.name});
    }
};

pub fn main() void {
    var dog = std.heap.c_allocator.create(Dog) catch unreachable;
    dog.* = Dog{ .base = Animal{ .kind = .dog }, .name = "Ziggy" };

    // this works fine, but only because it's *Dog
    std.debug.print("{}\n", .{dog});

    // this should still work, but it segfaults
    var animal = &dog.base;
    std.debug.print("{}\n", .{animal});
}
```

The problem here is that `std.fmt.formatType()` dereferences the value before calling `val.format()` which will turn it into `*const T`, and then the polymorphic of course `@ptrCast` cannot work, because it's a pointer to a copy of `Animal`, not the whole `Dog`.

The problem is that all of this was completely not obvious, and I was not getting any meaningful errors along the way.

Also, I'd argue (but this is subjective) that it should be possible to format pointer types, without making a copy.

![image](https://user-images.githubusercontent.com/3526922/232416332-86f14e99-e785-4788-a47c-871dcce30de2.png)

### Expected Behavior

- It should not segfault.
- (Optional) There should be a way to provide `format()` for pointer types. Currently, it only works for `*const` pointers, and it's just by accident because it's not using the original pointer at all.

**Comment by dvmason** [comment]
Similar problem.... why does formatting make a copy of the object? Obviously broken if the struct is variable sized.

See example here.... the address in the `format` is not the same as the address in the caller. And if I pass a pointer to a larger struct,  it copies the original (not surprising, if you're going to deref and copy.... but why would you do  that? What happened to "no surprises"?!?!!?

See this code: https://godbolt.org/z/131W34jfq

**Comment by dvmason** [comment]
> BTW: I'd be happy to help, and I also think that it should be possible to make it work without actually breaking anything if we restrict the solution just to `*const T` types.

I looked at the code (in ZIG/lib/std/fmt.zig) and formatType looks like it could be called without the dereference. Pretty easy would be to split off all the code after the check for a builtin `format` function for the type into a separate function and call *that* with a dereference. But there might be a better way that would avoid the copy entirely. (And the `const` part isn't relevant.)

I'd try to fix it, but I'm on a tight deadline, so I just made a `write` function which has all the code from my `format`, and that works for what I need right now.
